Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp

```diff
@@ -0,0 +1,294 @@
+        return;
+    }
+
+    spiel_speaker_new(nullptr, [](GObject*, GAsyncResult* result, gpointer userData) {
+        auto self = reinterpret_cast<SpielSpeechWrapper*>(userData);
+        GUniqueOutPtr<GError> error;
+        self->m_speaker = adoptGRef(spiel_speaker_new_finish(result, &error.outPtr()));
+        if (error) {
+            WTFLogAlways("Spiel speaker failed to initialize: %s", error->message);
+            return;
+        }
+        self->finishSpeakerInitialization();
+    }, this);
+}
+
+void SpielSpeechWrapper::finishSpeakerInitialization()
+{
+    g_object_set(m_speaker.get(), "sink", m_sink.get(), nullptr);
+
+    // TODO: Plumb support for boundaryEventOccurred? Using range-started signal?
+
+    g_signal_connect_swapped(m_speaker.get(), "utterance-started", G_CALLBACK(+[](SpielSpeechWrapper* self, SpielUtterance*) {
+        self->m_platformSynthesizer.client().didStartSpeaking(*self->m_utterance);
+    }), this);
+
+    g_signal_connect_swapped(m_speaker.get(), "utterance-finished", G_CALLBACK(+[](SpielSpeechWrapper* self, SpielUtterance*) {
+        self->m_platformSynthesizer.client().didFinishSpeaking(*self->m_utterance);
+        self->clearUtterance();
+    }), this);
+
+    g_signal_connect_swapped(m_speaker.get(), "utterance-canceled", G_CALLBACK(+[](SpielSpeechWrapper* self, SpielUtterance*) {
+        self->m_platformSynthesizer.client().didFinishSpeaking(*self->m_utterance);
+        self->clearUtterance();
+    }), this);
+
+    g_signal_connect_swapped(m_speaker.get(), "utterance-error", G_CALLBACK(+[](SpielSpeechWrapper* self, SpielUtterance*) {
+        self->m_platformSynthesizer.client().speakingErrorOccurred(*self->m_utterance);
+        self->clearUtterance();
+    }), this);
+
+    g_signal_connect_swapped(m_speaker.get(), "notify::paused", G_CALLBACK(+[](SpielSpeechWrapper* self, SpielSpeaker* speaker) {
+        gboolean isPaused;
+        g_object_get(speaker, "paused", &isPaused, nullptr);
+        if (isPaused)
+            self->m_platformSynthesizer.client().didPauseSpeaking(*self->m_utterance);
+        else
+            self->m_platformSynthesizer.client().didResumeSpeaking(*self->m_utterance);
+    }), this);
+
+    g_signal_connect_swapped(m_speaker.get(), "notify::voices", G_CALLBACK(+[](SpielSpeechWrapper* self, SpielSpeaker*) {
+        self->m_platformSynthesizer.client().voicesDidChange();
+    }), this);
+
+    m_speakerCreatedCallback();
+}
+
+SpielSpeechWrapper::~SpielSpeechWrapper()
+{
+    g_signal_handlers_disconnect_by_data(m_speaker.get(), this);
+}
+
+String SpielSpeechWrapper::generateVoiceURI(const GRefPtr<SpielVoice>& voice, const String& language)
+{
+    auto provider = spiel_voice_get_provider(voice.get());
+    return makeString(URI_PREFIX, span(spiel_provider_get_well_known_name(provider)), '#', span(spiel_voice_get_identifier(voice.get())), '#', language);
+}
+
+Vector<RefPtr<PlatformSpeechSynthesisVoice>> SpielSpeechWrapper::initializeVoiceList()
```

Comment:
I don't know if it makes sense or not but I think you could split this into two methods:

* private initialize that stores the voices in an attribute.
* `voiceList` that returns `const Vector...  &` of that attribute.

Reply:
I don't think that's needed :)